README
$Id: README,v 1.23 2008/10/02 18:53:54 turnstep Exp $

This is Net::SSH::Perl. It contains implementations of
both the SSH1 and SSH2 protocols.

PREREQUISITES

Protocol 1

    * Math::GMP (1.04 or greater)
    * String::CRC32 (1.2 or greater)
    * Digest::MD5
    * IO::Socket
    * Crypt::IDEA

Protocol 2

    * CryptX (0.032 or greater)
    * Crypt::Curve25519 (0.05 or greater)
    * IO::Socket

It optionally requires Digest::BubbleBabble for
generating bubble babble fingerprints.

INSTALLATION

Net::SSH::Perl installation is relatively straightforward. If
your CPAN shell is set up, you should just be able to do

    % perl -MCPAN -e 'install Net::SSH::Perl'

If you don't like that, you can download the distribution; the
latest version on CPAN can be found at

	http://search.cpan.org/dist/Net-SSH-Perl/

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

SAMPLES/TUTORIALS

Take a look at the scripts in eg/ for help and examples of
using Net::SSH::Perl.

* eg/cmd.pl is just a simple example of some of the
functionality in the module;

# eg/pssh is an ssh-like client for running commands on
remote servers and connecting to a shell on remote servers;

* eg/pscp is a very simple scp-like script;

* eg/pssh-keygen is, as you might guess, an implementation of
ssh-keygen in Perl, using the Net::SSH::Perl key libraries.
It contains almost all of the functionality in the OpenSSH
version of ssh-keygen.

* eg/remoteinteract.pl is an example of using register_handler
to interact with a remote (interactive) command.

Both pssh and pscp support a subset of the command line options
that the actual tools support; obviously, only those options
supported by Net::SSH::Perl are supported by the two scripts.

David Robins / dbrobins@cpan.org

Support for the following features by:
Lance Kinley / lkinley@loyaltymethods.com

Diffie-Hellman Group 14, 16, and 18 Key Exchange
Diffie-Hellman Group Exchange
Curve25519 Key Exchange (curve25519-sha256@libssh.org)
AES Cipher support (aes256-ctr,aes192-ctr,aes128-ctr)
Chacha20-Poly1305 cipher support (chacha20-poly1305@openssh.com)
hmac-sha2-256,hmac-sha2-512 MAC support
hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com
  Encrypt-then-MAC (ETM) MAC support
Ed25519 key support, including encrypted keys that are enciphered
  with supported crypto algorithms
ECDSA Elliptical Curve DSA key support: ecdsa-sha2-nistp256,
  ecdsa-sha2-nistp384, ecdsa-sha2-nistp521


Changes
Revision history for Net::SSH::Perl

2.13 2017.06.30
    - Bug Fix: AES-CBC IV length needs to be blocksize, not keysize
    - Bug Fix: DES3 init needs to pass key as variable to avoid scalar error
    - Bug Fix: Packet padlen decode was using signed 8-bit unpack (should be unsigned)
    - Fix CBC inefficiency
    - Fix Key not being loaded prior to Agent use [ https://github.com/lkinley/Net-SSH-Perl/issues/10 ]

2.12 2017.04.22
    - Fix inefficiency in CTR, revealed by profiling with Devel::NYTProf
    - Avoid warnings in Host from strange/invalid known_host entries
    - Improve documentation of newer features

2.11 2017.04.16
    - Packet bugfix: Introduced in 2.07, when ETM Mac is used, sometimes
      not enough incoming bytes were available to fully read stored MAC
    - DSA key bugfix: verify would fail if r/s had highest bit set

2.10 2017.03.23
    - Add curve25519-sha256 alias for curve25519-sha256@libssh.org key exchange
    - Bugfix for when unsupported key types are encountered
    - Add support for '-' syntax in options, including wildcards
    - Add wildcard support for '+' syntax in options

2.09 2016.10.26
    - Fix creation of keys in ecdsa, ed25519 key classes
    - Update eg/pssh-keygen to create ecdsa, ed25519 keys
    - Handle hostkeys-00@openssh.com global requests
    - Add support for 'CheckHostIP' and 'UpdateHostKeys' config options
    - Refactor handling of '+' syntax in options
    - Key fingerprints now output sha256-base64 by default.
      (md5 can be specified with FingerprintHash config option)
    - Add id_ed25519, id_ecdsa to default identity files
    - Documentation updates in Perl.pm to reflect new functionality in 2.XX

2.08 2016.10.14
    - Use sha512 instead of md5 in Net::SSH::Perl::Cipher->new_from_key_str()
      to provide ChachaPoly with enough key material
      Tests in t/05-cipher.t should now pass on all platforms [ CPAN bug #114077 ]
    - Add AES128_CBC to cipher tests
    - Info on using features not enabled by default added to README

2.07 2016.10.13

    - Fix blowfish compilation on SunOS [CPAN bug #116323]
    - Fix bug in Packet [CPAN bug #118335]
    - Add support for '+' syntax in MACs option
    - Remove hmac-sha1 from default MACs. It can re-enabled 
      by passing the option: 'MACs +hmac-sha1'

2.06 2016.10.04

    - Add support for additional fixed Diffie-Hellman 2K, 4K and 8K groups
      from OpenSSH 7.3 (draft-ietf-curdle-ssh-kex-sha2-03)
    - Kex defaults now updated to draft-ietf-curdle-ssh-kex-sha2-03
      recommendations (diffie-hellman-group-exchange-sha1 removed)
      It can re-enabled by passing the option:
      'KexAlgorithms +diffie-hellman-group-exchange-sha1'

2.05 2016.10.03

    - Add support for '+' syntax in Ciphers, KexAlgorithms, HostKeyAlgorithms
      options as in OpenSSH

2.04 2016.05.11

    - Add ECDSA key support
    - Improve extract_public() in Key.pm inspired by
      https://github.com/renormalist/Net-SSH-Perl/pull/12
      but implement comment with backwards compat with RSA/DSA datafellows
    - Fix XS from being loaded more than once (warnings from Net::SFTP)

2.03 2016.05.06

    - Fixes so that "make test" passes

2.02 2016.05.04

    - Use CryptX to further reduce module depedencies
      This eliminates the need for:
        Math::Pari
        Crypt::DH
        Crypt::RSA
        Crypt::DSA
        Crypt::DES
        Crypt::Blowfish
        MIME::Base64
    - Add support for rsa-sha2-512,rsa-sha2-256 signing with RSA keys
    - Implement HashKnownHosts, KexAlgorithms, MACs config directives
    - Add XS code for Chacha20, BSD Blowfish, Ed25519 routines
    - Properly handle and create known_hosts entries when port is specified
    - Remove obsolete ciphers, MACs, Kex from default list to duplicate 
      upcoming OpenSSH behavior
    - Bug fixes

2.01 2016.02.19

    - Use CryptX to reduce module depedencies
      This eliminates the need for:
        BSD::arc4random
        Digest::MD5
        Digest::SHA
        Digest::HMAC_MD5
	Crypt::OpenSSL::AES

2.00 2015.12.07

    - Add Chacha20-Poly1305 cipher support for best security
      (Requires Crypt::OpenSSH::ChachaPoly, see README)
    - Add AES Cipher support in CTR mode (CBC mode supported in Ed25519
      keys only)
    - Add Group Exchange (RFC4523) Diffie-Hellman Key Exchange
    - Add Curve25519 (curve25519-sha256@libssh.org) Key Exchange support
      (Requires Crypt::Curve25519)
    - Add hmac-sha2-256,hmac-sha2-512 MAC support
    - Add hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com
      Encrypt-then-MAC (ETM) MAC support
    - Use BSD::arc4random for encrypted packet padding
    - Add support for Ed25519 ssh/host keys (Requires Crypt::Ed25519)
      Encrypted Ed25519 key support requires Crypt::OpenBSD::Blowfish
      (See README for info)
    - Default ciphers order is now chacha,aes,3des,blowfish,arcfour
    - Default KEX order is now Curve25519, DHGEXSHA256, DHGEXSHA1, DH14, DH1
    - Default MAC order is now hmac-sha2-512-etm@openssh.com,
      hmac-sha2-256-etm@openssh.com, sha2-512, sha2-256, sha1, md5
    - SSH Keys can now be in DOS format (no need to remove CR/LF)
    - SOCKS proxy support via sub class Net::SSH:Perl::Proxy
    - Now does not abort due to OpenSSH 6.8+ server
      SSH2_MSG_GLOBAL_REQUEST messages for host key rotation

1.42 2015.09.28
    - Fix issue with long selects getting interrupted by signals and dying (Andrew Hoying)
    - fix version cpan meta info

1.41 2015.09.18
    - use Errno constants in more portable way, see perldoc Errno

1.40 2015.09.17
    - declare new dependency to File::HomeDir in Makefile.PL

1.39 2015.09.15

    - RT #83978 - fix shell terminal width and height (lharey)
    - RT #94574 - fix Algorithm negotiation issue in ::Kex.pm
      (Brian Curnow, Michael Gray)
    - RT #105728 - fix VERSION methods (reported by Karen Etheridge)
    - code modernization (strict+warnings) (gh0stwizard)
    - pass tests on Win32 (gh0stwizard)

Makefile.PL
# $Id: Makefile.PL,v 1.30 2008/10/02 18:53:25 turnstep Exp $

use strict;
use 5.006;
use ExtUtils::MakeMaker qw(prompt WriteMakefile);

my %prereq = (
        'IO::Socket'        => 0,
        'File::Spec'        => 0,
        'File::HomeDir'     => 0,
);

my %SSH_PREREQ = (
    1 => {
        'Digest::MD5'       => 0,
        'String::CRC32'     => '1.2',
        'Math::GMP'         => '1.04',
        'Scalar::Util'      => 0,
    },

    2 => {
        'CryptX'            => '0.032',
        'Crypt::Curve25519' => '0.05',
        'Convert::PEM'      => '0.05',
    },
);

$SSH_PREREQ{3} = { map %{$SSH_PREREQ{$_}}, 1..2 };

print<<MSG;
This is Net::SSH::Perl.

As of version 1.00, Net::SSH::Perl supports both the SSH1 and
SSH2 protocols natively. The two protocols have different
module prerequisitives, so you need to decide which protocol(s)
you plan to use. If you use one or the other, only those modules
for your chosen protocol will be installed; if you choose both,
all of the supporting modules will be installed. Please choose
the protocols you'd like to use from the following list ("Both"
is the default).

MSG

printf "    [%d] SSH%d\n", $_, $_ for 1..2;
printf "    [3] Both SSH1 and SSH2\n";

my $p = prompt("\nWhich protocol(s) do you plan to use?", 3);
print "\n";

@prereq{keys %{$SSH_PREREQ{$p}}} = values %{$SSH_PREREQ{$p}};

print "Checking for optional modules\n\n";

unless (have_module('Digest::BubbleBabble', 0.01)) {
    print<<MSG, "\n";
Digest::BubbleBabble is required if you want to generate bubble babble
key fingerprints with pssh-keygen.
MSG

    if (read_yes_or_no("Would you like to install it? (y/n)", "y")) {
        $prereq{'Digest::BubbleBabble'} = '0.01';
    }
    print "\n";
}

WriteMakefile(
    (MM->can('signature_target') ? (SIGN => 1) : ()),
    NAME         => 'Net::SSH::Perl',
    DISTNAME     => 'Net-SSH-Perl',
    VERSION_FROM => 'lib/Net/SSH/Perl.pm',
    PREREQ_PM    => \%prereq,
    AUTHOR       => 'David Robins <dbrobins@cpan.org>',
    ABSTRACT     => 'Perl client interface to SSH',
    NO_META      => 1,
);

sub read_yes_or_no {
    my($prompt, $def) = @_;
    my $ans = prompt($prompt, $def);
    $ans =~ /^y/i;
}

sub have_module {
    my($name, $ver) = @_;
    eval("use $name" . ($ver ? " $ver;" : ";"));
    !$@;
}
